### ISSUE TITLE
Auto-converting 'let' to 'const' can break code in some situations

### ISSUE THREAD

**Issue by fenomas** [standard/standard#1765]
**What version of this package are you using?**
standard@16.0.4

**What operating system, Node.js, and npm version?**
the [online repl](https://standardjs.com/demo.html#try-it-out)

**What happened?**
Apps that autoformat via standard sometimes break code by converting `let` statements to `const` if the variable in question is never reassigned.

I mainly see this in two situations:

 * When a variable is declared with `let`, and later reassigned, but the reassignments are temporarily commented out (for debugging, etc)
 * When a variable is *intended* to be reassignable, and is correctly declared with `let` to convey that intention, but the reassignments aren't in place yet (because the file is a stub, template, tutorial, etc.)

In either case, standard/fix will convert the `let` to `const` and break the code (in the sense that the code will break in the future when the reassignable variable gets reassigned).

**What did you expect to happen?**
A warning about `let` variables that are not reassigned would be the right approach, without auto-converting them.

**Are you willing to submit a pull request to fix this bug?**
No

**Comment by voxpelli** [comment]
In my personal opinion:

This same argument can be applied to lots of in-progress code.

Comment out an `if`-statement and you'll get complaints about the indentation of its content.

A linter as I see it is supposed to protect and improve the final product, not the in-progress code.

If one eg. have an auto-fix on save, then the easiest fix is to remove that auto-fix and instead do manual fixing.

I personally never run any automated fixing and often opt to fix things manually, by eg using the hover boxes that appear on lint errors in VS Code.

**Comment by voxpelli** [comment]
@fenomas I can see the issue with this rule when it comes to autofixing on save, but I'm not convinced that is necessarily an issue with `standard`.

If we see it as a design goal that `standard` should work painlessly with autofix on save, then yes, I agree, this is a bug.

Designing `standard` to work well with autofix on save would make it less powerful in other scenarios though.

How have you set up your autofix on save? Using the `standard` vscode extension? (I'm not terribly familiar with that extension, but I think eg. @Divlo knows it quite well)

You could set up your own `eslint` setup with `eslint-config-standard` and override the rule for `let`/`const` there if you want to get around it for now.

**Comment by voxpelli** [comment]
Rule in question: [`prefer-const`](https://eslint.org/docs/rules/prefer-const#suggest-using-const-prefer-const)

To note is also that there's not really any way of having a rule active in ESLint, but disabling the autofix for it. So you would have to make it [on the call to the fix](https://stackoverflow.com/a/62949733):

```bash
eslint somedir --fix --rule "somerule: 0"
```

Whether something like that would be possible to add to eg. the `standard` extension for VSCode, that I don't know unfortunately.

**Comment by fenomas** [comment]
> If we see it as a design goal that standard should work painlessly with autofix on save, then yes, I agree, this is a bug.  
> Designing standard to work well with autofix on save would make it less powerful in other scenarios though.

@voxpelli This is a confusing answer, as autofix-on-save appears to be a configurable option in all of the official editor plugins. All of their docs have a note about how to enable it, with no warnings - even the readme of this repo has a note about it. It looks like a regular bread-and-butter feature, not an unsupported (much less unsupportable) edge case.

With that said though, personally I don't use fix-on-save, and I don't think it's directly related to the issue. The high-order bit is, autoformat is normally a "safe", non-destructive operation. Nobody assumes that it will always fix *all* errors, but people do assume that it won't change working code into a non-working state. 

So personally, I use it more frequently than I save - whenever something needs to be formatted, I hit autoformat. And I can't recall ever having seen an autoformatter that wasn't safe by default - not with standard (outside of this issue), prettier, eslint, vscode's default formatter, etc.

**Comment by Divlo** [comment]
Just to be clear: We use the formatting behavior of ESLint.
And the same thing for the auto-fix thing in the VSCode extension.

Keep in mind that `standard` is only a wrapper around ESLint to get linting quickly but under the hood, it uses the APIs ESLint provides.

If you really want this behavior, you could open an issue in the ESLint repo.
But as said:
> A linter as I see it is supposed to protect and improve the final product, not the in-progress code.

**Comment by LinusU** [comment]
> [...] but people do assume that it won't change working code into a non-working state.

I have yet to se it demonstrated that it turns any code into a non-working state though?

**Comment by fenomas** [comment]
> We use the formatting behavior of ESLint.

I know, and eslint's formatter is safe by default. The rule we're talking about is not part of `eslint:recommended`.

**Comment by voxpelli** [comment]
> > A linter as I see it is supposed to protect and improve the final product, not the in-progress code.
> 
> I don't know what that means. When is code ever "final"?

A linter looks at the piece of code as is. It doesn't assume that any code has been commented out or should be expected to be added. It looks at the code as is and sees if it fulfils the requirements that has been set up.

`standard` has a requirement that `const` should be used wherever it can be used.

That _"can"_ is determined by looking at the state of the project as its linted.

If you run `standard` without ` --fix`, then you'll get errors and warnings. If you run `standard --fix`, then it will fix your current state of the project to adhere to as many of those requirements as possible.

**Comment by fenomas** [comment]
> A linter looks at the piece of code as is. It doesn't assume that any code has been commented out or should be expected to be added.

I fully agree, but that's the opposite of what's been said so far. If the linter requires code to be "final" (still no idea what that means), then it's obviously making assumptions about what future edits will (not) happen. If it takes code as-is, then it shouldn't assume anything about the state of the code.

I mean, consider the error where a variable is declared but never used. If the code is "final", then it's safe to fix that error by removing the variable entirely. But you wouldn't want your formatter to remove all unused variables silently by default, right? I'm saying the same logic should apply here.

----

**PS:** is this "linters are only supposed to work with final code" idea discussed anywhere I could look at? I've used eslint and other linters pretty extensively and I've never heard anything at all similar.

**Comment by dougwilson** [comment]
Hi @fenomas the `standard` is based on `eslint` directly. For your above comparison, the `eslint` `no-unused-vars` is not a fixable rule, so that is why `--fix` does not remove them, but `eslint` has the `prefer-const` rule as fixable, so `--fix` will change your `let`s into `const`s when able. It is actually `eslint`, not `standard` that is making the distinction upon which one of those are changed when you run `--fix`.

**Comment by fenomas** [comment]
@dougwilson Of course I'm aware that `unused-vars` isn't fixable in eslint - I was just making an illustrative comparison for why the `prefer-const` rule shouldn't be applied by default.

> It is actually eslint, not standard that is making the distinction upon which one of those are changed

It's standard that's choosing which fixable rules to apply. ESlint has a number of rules that are fixable, but which standard does not autofix by default - I'm suggesting that `prefer-const` should be one of them.

**Comment by dougwilson** [comment]
So just to clarify: are you saying that standard should never make suggestions for if vars should be `const` or not at all, or that it should, just that it shouldn't auto fix them?

**Comment by fenomas** [comment]
@dougwilson 
> **What did you expect to happen?**
> A warning about let variables that are not reassigned would be the right approach, without auto-converting them.

But if that's not possible, and the only options are both warning+autofix, or neither, then neither would seem preferable. At the end of the day, there are situations where an author declares a `let` intentionally, to convey that a variable *may be* reassigned even though it isn't reassigned currently. In those cases, autoconverting to const throws away the developer's semantic intent.

**Comment by dougwilson** [comment]
Right, that is what I'm trying to say: that a warning without autofix is not possible, at least not working you opening a feature request for that in eslint, as eslint doesn't offer such a configuration for this. That is why the above was pointing you to eslint, if that makes sense.

**Comment by fenomas** [comment]
Thanks for clarifying that @dougwilson. In that case, per my last comment I think the right thing to do here is for standard to not enable `prefer-const` by default.

PS: it would be helpful if you could shed any light on the "linters are only for the final product, not the in-progress code" policy quoted extensively above. I've looked around various eslint docs, and the policy doesn't appear to exist outside this thread.

**Comment by dougwilson** [comment]
I am not familiar with that policy, and unfortunately cannot shed any light on it.

**Comment by fenomas** [comment]
I only ask because (a) the policy appears to not exist, and (b) it's the only reason given so far for why this issue is closed.

Honestly, I'm not trying to be a weird hardass here - if the answer to this issue is "we agree there are potential risks but standard is opinionated and our opinion is to keep the current behavior", then that's 100% a reasonable answer. But currently the stated answer here is "fixing on save isn't supported because linters are only meant for final code", and that's not a very reasonable answer if neither standard nor eslint actually has such a policy.

**Comment by dougwilson** [comment]
Hi @fenomas apologies on not being able to elaborate; I am actually just a user of standard (and so somewhate vested in what the rules are) but not technically part of the project as a maintainer or anything. Apologies if I caused any confusion on that. You can typically see what role different commenters have based on the badge github places on the comments, though.

**Comment by fenomas** [comment]
@dougwilson Understood, and thanks for clarifying.

**Comment by voxpelli** [comment]
> it would be helpful if you could shed any light on the "linters are only for the final product, not the in-progress code" policy quoted extensively above

There is no such "policy" and it's also not something where `standard` differs from other linters.

(As eg. pointed out, the decision to make `prefer-const` fixable lies with ESLint)

So, no policy, just me trying to formulate my thoughts.

If you want an explicit reason for the closing this then I would eventually close this due to:

* The decision to make `prefer-const` fixable lies with ESLint
* `standard` should not remove `prefer-const`

Hence making this an issue with ESLint, not `standard`.

**Comment by fenomas** [comment]
> The decision to make prefer-const fixable lies with ESLint  
> Hence making this an issue with ESLint, not standard.

@voxpelli I'm not suggesting that `prefer-const` should not be fixable. I'm saying it should not be applied by default. 

ESLint does not apply the rule by default, so there is no issue to report to ESLint. Standard does apply it by default, so I'm reporting here. Thanks for understanding.

**Comment by voxpelli** [comment]
Right, here's the discussions why it was added:

* https://github.com/standard/eslint-config-standard/pull/133
* https://github.com/standard/standard/issues/523

**Comment by fenomas** [comment]
Thank you for the links - I searched issues, but I didn't think to search pull requests.

**Comment by fenomas** [comment]
I've just realized, maybe the disconnect here is whether `standard --fix` is an **autoformat** command, or whether it's a **fix conformance errors** command.

The readme and other docs here describe it as an autoformatter, so that's what I've been assuming. And if you use `standard --fix` like an autoformatter (meaning you run it frequently, whenever something isn't formatted correctly) then sooner or later `prefer-const` inevitably breaks your code. The more frequently you autoformat, the more it will happen.

But it sounds like maintainers here are using `--fix` more as an occasional conformance tool, which is only run at specific times (e.g. as a hook before tests or commits). If used that way, I can certainly see how `prefer-const` wouldn't cause any problems.

@voxpelli, is that latter usage what you meant about using `--fix` on final code? If so then I think we were talking past each other, as I've been assuming the former kind of usage.

**Comment by voxpelli** [comment]
@fenomas **_Very_** good formulation of the situation, thanks! 🙏 

You are entirely correct. To me it's a `fix conformance errors` command.

@LinusU @Divlo What are your perceptions? Should we improve / clarify the docs on this?

**Comment by fenomas** [comment]
Thanks for confirming @voxpelli - apologies if I sounded snippy before, but I was really confused why it would be necessary to finalize code before autoformatting. 😅 I'll edit the original issue to make things clearer.

As for the issue itself, to be honest if `prefer-const` was removed then I think `standard --fix` can be pretty functionally used as a regular autoformatter. I haven't deeply investigated, but I was previously using it that way and I don't remember it breaking anything (outside of this issue).

OTOH if `--fix` isn't meant to be used frequently, documenting that also works. But that means `standard` users realistically need to use something else an autoformatter, which may lead to confusion if their formatter does anything that contradicts `--fix`.
